### PR TITLE
feat(vap): add --timeout flag to deploy-library command

### PR DESCRIPTION
**Pull request by manmathbh**
Adds a configurable `--timeout` flag to the `deploy-library` command.

## Changes
- `--timeout` flag accepts duration values (e.g. `30s`, `1m`), defaults to `0s` (no timeout)
- Replaced bare `http.Get()` with `http.Client{Timeout}` for configurable timeouts
- `deployLibrary()` and `downloadFileToString()` now propagate the timeout parameter

## Usage
```bash
kubescape vap deploy-library --timeout 30s | kubectl apply -f -
```

## Tests (85 total, 8 new for timeout)
- Flag registration and default value (`0s`)
- Cobra flag parsing (`--timeout 30s` → `30*time.Second`)
- Timeout `0` means no timeout (fast download)
- Short timeout (`10ms`) triggers error on slow server
- Non-zero timeout (`5s`) allows fast downloads

`go vet` clean.

## Summary by CodeRabbit

* **New Features**
  * `deploy-library` adds a timeout option and can write combined YAML to a file or stdout.
  * `create-policy-binding` adds `--output/-o` to write YAML to a file or stdout.

* **Bug Fixes**
  * Label selector validation now uses Kubernetes parsing and enforces equality selectors.
  * Kubernetes object name validation accepts longer names (up to 253 chars) with existing rules.

* **Tests**
  * Expanded tests for timeout-aware downloads, deploy behavior, validation, and deterministic CLI output.

### CODE REVIEW

**Comment by manmathbh** [comment]
PTAL @matthyx

File: cmd/vap/vap_test.go [hunk 1]
```diff
@@ -1,10 +1,651 @@
...
+		{name: "ends with hyphen", input: "abc-", wantErr: true, errMsg: "should consist of lower case alphanumeric characters"},
+
+		// invalid - uppercase
+		{name: "contains uppercase", input: "Abc", wantErr: true, errMsg: "should consist of lower case alphanumeric characters"},
+		{name: "all uppercase", input: "ABC", wantErr: true, errMsg: "should consist of lower case alphanumeric characters"},
+
+		// invalid - special characters
+		{name: "contains underscore", input: "abc_def", wantErr: true, errMsg: "should consist of lower case alphanumeric characters"},
+		{name: "contains space", input: "abc def", wantErr: true, errMsg: "should consist of lower case alphanumeric characters"},
+		{name: "contains dot in middle (not allowed by regex)", input: "abc.def", wantErr: true, errMsg: "should consist of lower case alphanumeric characters"},
```
Comment (matthyx) [inline cmd/vap/vap_test.go]: This test case locks in behavior that does not match Kubernetes object-name rules. Dotted DNS-subdomain names are valid metadata names, so `abc.def` should not be treated as invalid here. For example, `k8s.io/apimachinery/pkg/util/validation.IsDNS1123Subdomain("team.prod")` returns no errors.

As written, this test will preserve the current mismatch and make a future fix harder to land.

File: cmd/vap/vap_test.go [hunk 2]
```diff
@@ -1,10 +1,651 @@
...
+		{name: "simple key=val", input: "app=nginx", wantValid: true},
+		{name: "key and val with digits", input: "env1=prod2", wantValid: true},
+		{name: "uppercase allowed", input: "App=Value", wantValid: true},
+		{name: "mixed case", input: "appName=NginxValue", wantValid: true},
+		{name: "missing equals (space)", input: "key value", wantValid: false},
+		{name: "missing value", input: "key=", wantValid: false},
+		{name: "missing key", input: "=value", wantValid: false},
+		{name: "multiple equals", input: "key=val=extra", wantValid: false},
+		{name: "empty string", input: "", wantValid: false},
+		{name: "contains hyphen", input: "app-name=nginx", wantValid: false},
```
Comment (matthyx) [inline cmd/vap/vap_test.go]: These cases are asserting an invalid label-selector contract. Kubernetes accepts keys like `app-name`, `app.name`, `app_name`, and qualified names such as `app.kubernetes.io/name`; for example, `k8s.io/apimachinery/pkg/labels.Parse("app.kubernetes.io/name=nginx")` succeeds.

That means the test is hardening the current bug instead of catching it, so valid selectors would continue to be rejected by `create-policy-binding`.

**Comment by manmathbh** [comment]
@matthyx fixed both, `isValidK8sObjectName` now allows dots (abc.def, team.prod-v2) and uses the 253-char DNS subdomain limit.
Label regex also updated to accept hyphens/dots/underscores in keys and values.
Tests now go through `cmd.Execute()` instead of the old inline checks.
PTAL~

File: cmd/vap/vap.go [hunk 3]
```diff
@@ -81,7 +96,7 @@ func getCreatePolicyBindingCmd() *cobra.Command {
 			}
 			for _, label := range labelArr {
 				// Label selector must be in the format key=value
-				if !regexp.MustCompile(`^[a-zA-Z0-9]+=[a-zA-Z0-9]+$`).MatchString(label) {
+				if !regexp.MustCompile(`^[a-zA-Z0-9]([a-zA-Z0-9._-]*[a-zA-Z0-9])?=[a-zA-Z0-9]([a-zA-Z0-9._-]*[a-zA-Z0-9])?$`).MatchString(label) {
```
Comment (matthyx) [inline cmd/vap/vap.go]: This validation is still stricter than Kubernetes label-selector rules, so `create-policy-binding` rejects valid selectors such as `app.kubernetes.io/name=nginx`, `kubescape.io/infra=config`, and `key=`. `k8s.io/apimachinery/pkg/labels.Parse(...)` accepts all three.

Because `--label` is turned into `ObjectSelector.MatchLabels`, the CLI should validate with Kubernetes label key/value rules rather than a custom regex here.

**Comment by manmathbh** [comment]
@matthyx , swapped the regex for labels.Parse, handles the qualified names and empty values now. PTAL

File: cmd/vap/vap.go [hunk 4]
```diff
@@ -88,8 +96,7 @@ func getCreatePolicyBindingCmd() *cobra.Command {
 				}
 			}
 			for _, label := range labelArr {
-				// Label selector must be in the format key=value
-				if !regexp.MustCompile(`^[a-zA-Z0-9]+=[a-zA-Z0-9]+$`).MatchString(label) {
+				if _, err := labels.Parse(label); err != nil {
```
Comment (matthyx) [inline cmd/vap/vap.go]: Using `labels.Parse` here widens accepted input beyond the equality selectors that `createPolicyBinding` can actually encode. The later conversion still splits each label on `=` into `MatchLabels`, so valid selectors like `app` now panic (`index out of range` at `vap.go:250`), and selectors like `app!=nginx` are silently rewritten to invalid semantics (`matchLabels: {app!: nginx}`).

Please either translate the parsed selector into a full `metav1.LabelSelector` (`MatchLabels` + `MatchExpressions`) or restrict validation back to equality-only selectors until that translation exists.

File: cmd/vap/vap.go [hunk 5]
```diff
@@ -88,10 +97,16 @@ func getCreatePolicyBindingCmd() *cobra.Command {
...
 			for _, label := range labelArr {
-				// Label selector must be in the format key=value
-				if !regexp.MustCompile(`^[a-zA-Z0-9]+=[a-zA-Z0-9]+$`).MatchString(label) {
+				parsed, err := labels.Parse(label)
+				if err != nil {
 					return fmt.Errorf("invalid label selector: %s", label)
 				}
+				requirements, _ := parsed.Requirements()
+				for _, r := range requirements {
+					if r.Operator() != selection.Equals && r.Operator() != selection.DoubleEquals {
```
Comment (matthyx) [inline cmd/vap/vap.go]: Allowing `selection.DoubleEquals` here still breaks YAML generation later. `createPolicyBinding()` continues to split the raw selector string on the first `=`, so `--label 'app==nginx'` is accepted but rendered as:

```yaml
matchLabels:
  app: =nginx
```

That is not the same selector. Please either normalize `==` to a single `=` before building `MatchLabels`, or reject `==` and only allow the single-`=` form the command can actually encode.

File: cmd/vap/vap.go [hunk 6]
```diff
@@ -88,10 +97,16 @@ func getCreatePolicyBindingCmd() *cobra.Command {
 				}
 			}
 			for _, label := range labelArr {
-				// Label selector must be in the format key=value
-				if !regexp.MustCompile(`^[a-zA-Z0-9]+=[a-zA-Z0-9]+$`).MatchString(label) {
+				parsed, err := labels.Parse(label)
```
Comment (matthyx) [inline cmd/vap/vap.go:100]: `labels.Parse` accepts selectors with surrounding whitespace, so `--label 'app = nginx'` gets through this validation. But later `createPolicyBinding()` still rebuilds `MatchLabels` by splitting the raw string on `=`, which emits:

```yaml
matchLabels:
  'app ': ' nginx'
```

That changes the selector instead of preserving `app: nginx`. Since this command only supports equality selectors, please build `MatchLabels` from the parsed requirement (or at least trim/normalize before serializing) and add a test for the whitespace form, which is not covered today.

**Comment by manmathbh** [comment]
@matthyx , switched to building `MatchLabels` from the parsed requirements instead of splitting on =, so whitespace, qualified names etc all work. added a test for app = nginx → {app: nginx}.
PTAL~

File: cmd/vap/vap.go [hunk 7]
```diff
@@ -186,16 +203,26 @@ func downloadFileToString(url string) (string, error) {
...
+	if len(name) == 0 {
+		return errors.New("name should not be empty")
+	}
+	if strings.HasPrefix(name, ".") || strings.HasSuffix(name, ".") {
+		return errors.New("name should not start or end with '.'")
 	}
 
-	regex := regexp.MustCompile(`^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`)
-	if !regex.MatchString(name) {
+	if !regexp.MustCompile(`^[a-z0-9]([-a-z0-9.]*[a-z0-9])?$`).MatchString(name) {
```
Comment (matthyx) [inline cmd/vap/vap.go]: This helper now models generic DNS-subdomain names (dots allowed, max 253 chars), but `--namespace` above still reuses it. Kubernetes namespace names are stricter DNS labels, so values like `team.prod` or a 64-character namespace now pass local validation even though `k8s.io/apimachinery/pkg/api/validation.ValidateNamespaceName` rejects them.

That lets `create-policy-binding` emit selectors that can never match a real namespace. Please validate `--namespace` with namespace-specific rules and add regression tests for dotted and >63-character namespaces.

**Comment by manmathbh** [comment]
@matthyx , split namespace validation into its own function, DNS label rules (63 chars, no dots). `team.prod` and long names get rejected now, added tests.
PTAL~

File: cmd/vap/vap.go [hunk 8]
```diff
@@ -186,22 +203,45 @@ func downloadFileToString(url string) (string, error) {
...
```
Comment (matthyx) [inline cmd/vap/vap.go]: This regex now allows dotted names, but it also accepts invalid DNS-1123 subdomains such as `a..b`, `a-.b`, and `a.-b`. Those values currently pass for `--name`, `--policy`, and `--parameter-reference`, even though `k8s.io/apimachinery/pkg/util/validation.IsDNS1123Subdomain(...)` rejects them.

That means the CLI can emit manifests the API server will reject later. Please validate with the upstream DNS-1123 subdomain helper (or equivalent per-label validation) and add negative tests for these cases.

**Comment by manmathbh** [comment]
@matthyx ,  switched both validators to` k8s.io/apimachinery/pkg/util/validation`, IsDNS1123Subdomain for `names/policies/paramrefs` and IsDNS1123Label for namespaces. catches a..b, a-.b, a.-b and everything else the upstream validators reject.
PTAL~

File: cmd/vap/vap.go [hunk 9]
```diff
@@ -171,6 +195,14 @@ func downloadFileToString(url string) (string, error) {
 	return bodyString, nil
 }
 
+func writeOutput(content string, outputFile string) error {
+	if outputFile != "" {
+		return os.WriteFile(outputFile, []byte(content), 0644)
```
[the remark this answers is not in the document]
Comment (matthyx) [reply cmd/vap/vap.go:211]: @manmathbh this still applies

[the remark this answers is not in the document]
Comment (manmathbh) [reply cmd/vap/vap.go:211]: @matthyx  added `os.MkdirAll` so nested paths work now.

**Review by matthyx — APPROVED** [review]
Rechecked on the latest head. The earlier blockers are resolved, and I do not see any remaining actionable issues in the current diff.
